File: packages/health_connector_hc_android/android/src/test/kotlin/com/phamtunglam/health_connector_hc_android/unit_tests/mappers/health_measurement_unit_mappers/PowerMapperTest.kt
package com.phamtunglam.health_connector_hc_android.unit_tests.mappers.health_measurement_unit_mappers

import androidx.health.connect.client.units.Power
import com.phamtunglam.health_connector_hc_android.mappers.health_measurement_unit_mappers.toDto
import com.phamtunglam.health_connector_hc_android.mappers.health_measurement_unit_mappers.toHealthConnect
import com.phamtunglam.health_connector_hc_android.pigeon.PowerDto
import com.phamtunglam.health_connector_hc_android.pigeon.PowerUnitDto
import io.kotest.matchers.shouldBe
import org.junit.jupiter.api.DisplayName
import org.junit.jupiter.api.Test

/**
 * Unit tests for Power Mapper.
 *
 * Tests verify proper bidirectional mapping between [PowerDto] and Health Connect
 * [Power] objects.
 */
@DisplayName("PowerMapper")
class PowerMapperTest {

    @Test
    @DisplayName(
        "GIVEN PowerDto in watts → " +
            "WHEN toHealthConnect called → " +
            "THEN creates Power with correct value",
    )
    fun whenPowerDtoInWatts_thenCreatesCorrectPower() {
        // Given
        val dto = PowerDto(value = 250.0, unit = PowerUnitDto.WATTS)

        // When
        val result = dto.toHealthConnect()

        // Then
        result.inWatts shouldBe 250.0
    }

    @Test
    @DisplayName(
        "GIVEN Power object → " +
            "WHEN toDto called → " +
            "THEN converts to DTO in watts",
    )
    fun whenPowerToDto_thenUsesWatts() {
        // Given
        val power = Power.watts(250.0)

        // When
        val result = power.toDto()

        // Then
        result.value shouldBe 250.0
        result.unit shouldBe PowerUnitDto.WATTS
    }
}

File: packages/health_connector_hc_android/test/unit_tests/src/mappers/measurement_unit_mappers/power_mapper_test.dart
import 'package:health_connector_core/health_connector_core_internal.dart';
import 'package:health_connector_hc_android/src/mappers/measurement_unit_mappers/power_mapper.dart';
import 'package:health_connector_hc_android/src/pigeon/health_connector_hc_android_api.g.dart';

import 'package:test/test.dart';

void main() {
  group(
    'PowerMapper',
    () {
      group(
        'PowerToDto',
        () {
          test(
            'converts Power to PowerDto in watts',
            () {
              const power = Power.watts(250.0);
              final dto = power.toDto();

              expect(dto.watts, 250.0);
            },
          );

          test(
            'converts Power from kilowatts to watts',
            () {
              const power = Power.kilowatts(0.25);
              final dto = power.toDto();

              expect(dto.watts, 250.0);
            },
          );
        },
      );

      group(
        'PowerDtoToDomain',
        () {
          test(
            'maps PowerDto to Power',
            () {
              final dto = PowerDto(watts: 250.0);
              final power = dto.toDomain();

              expect(power.inWatts, 250.0);
            },
          );
        },
      );
    },
  );
}
